crates/sui-core/src/overload_monitor.rs

```diff
@@ -0,0 +1,292 @@
+) {
+    info!("Starting system overload monitor.");
+
+    loop {
+        let authority_exist = check_authority_overload(&authority_state, &config);
+        if !authority_exist {
+            // `authority_state` doesn't exist anymore. Quit overload monitor.
+            break;
+        }
+        sleep(config.overload_monitor_interval).await;
+    }
+
+    info!("Shut down system overload monitor.");
+}
+
+// Checks authority overload signals, and updates authority's `overload_info`.
+// Returns whether the authority state exists.
+fn check_authority_overload(
+    authority_state: &Weak<AuthorityState>,
+    config: &OverloadThresholdConfig,
+) -> bool {
+    let authority_arc = authority_state.upgrade();
+    if authority_arc.is_none() {
+        // `authority_state` doesn't exist anymore.
+        return false;
+    }
+
+    let authority = authority_arc.unwrap();
+    let queueing_latency = authority
+        .metrics
+        .execution_queueing_latency
+        .latency()
+        .unwrap_or_default();
+    let txn_ready_rate = authority.metrics.txn_ready_rate_tracker.lock().rate();
+    let execution_rate = authority.metrics.execution_rate_tracker.lock().rate();
+
+    let (is_overload, load_shedding_percentage) =
+        check_overload_signals(config, queueing_latency, txn_ready_rate, execution_rate);
+    if is_overload {
+        authority
+            .overload_info
+            .set_overload(load_shedding_percentage);
+    } else {
+        authority.overload_info.clear_overload();
+    }
+
+    authority
+        .metrics
+        .authority_overload_status
+        .set(is_overload as i64);
+    authority
+        .metrics
+        .authority_load_shedding_percentage
+        .set(load_shedding_percentage as i64);
+    true
+}
+
+// Calculates the percentage of transactions to drop in order to reduce execution queue.
+// Returns the integer percentage between 0 and 100.
+fn calculate_load_shedding_percentage(txn_ready_rate: f64, execution_rate: f64) -> u32 {
+    // When transaction ready rate is practically 0, we aren't adding more load to the
+    // execution driver, so no shedding.
+    // TODO: consensus handler or transaction manager can also be overloaded.
+    if txn_ready_rate < 1e-10 {
+        return 0;
+    }
+
+    // Deflate the execution rate to account for the case that execution_rate is close to
+    // txn_ready_rate.
+    if execution_rate * 0.9 > txn_ready_rate {
+        return 0;
+    }
+
+    // In order to maintain execution queue length, we need to drop at least (1 - executionRate / readyRate).
+    // TO reduce the queue length, here we add 10% more transactions to drop.
+    (((1.0 - execution_rate * 0.9 / txn_ready_rate) + 0.1).min(1.0) * 100.0).round() as u32
+}
+
+// Given overload signals (`queueing_latency`, `txn_ready_rate`, `execution_rate`), return whether
+// the authority server should enter load shedding mode, and how much percentage of transactions to drop.
+fn check_overload_signals(
```

Comment:
One concern I have is, suppose for some reason we have a queueing delay that is unrelated to load. I can't imagine what would cause this, but perhaps we add such a bug in the future. It seems that in this case, the system would never recover no matter how much load we shed, so we would simply enter a death spiral. Is there anything we can do about that?

Reply:
Yes, the `max_load_shedding_percentage` in the configuration is our big red button. When such thing happens, we can turn off this feature by setting `max_load_shedding_percentage` to 0.